LLama/Grammar/GrammarParser.cs

```diff
@@ -0,0 +1,328 @@
+﻿using LLama.Native;
+using System;
+using System.Collections.Generic;
+
+namespace LLama.Grammar
+{
+    /// <summary>
+    /// Source:
+    /// https://github.com/ggerganov/llama.cpp/blob/6381d4e110bd0ec02843a60bbeb8b6fc37a9ace9/common/grammar-parser.cpp
+    /// 
+    /// The commit hash from URL is the actual commit hash that reflects current C# code.
+    /// </summary>
+    internal class GrammarParser
+    {
+        // NOTE: assumes valid utf8 (but checks for overrun)
+        // copied from llama.cpp
+        public Tuple<uint, ReadOnlyMemory<char>> DecodeUTF8(ReadOnlyMemory<char> src)
+        {
+            ReadOnlySpan<char> span = src.Span;
+            int[] lookup = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
+
+            byte firstByte = (byte)span[0];
+            byte highbits = (byte)(firstByte >> 4);
+            int len = lookup[highbits];
+            byte mask = (byte)((1 << (8 - len)) - 1);
+            uint value = (uint)(firstByte & mask);
+
+            int end = len;
+            int pos = 1;
+
+            for (; pos < end && pos < src.Length; pos++)
+            {
+                value = (uint)((value << 6) + ((byte)span[pos] & 0x3F));
+            }
+
+            ReadOnlyMemory<char> nextSpan = src.Slice(pos);
+
+            return new Tuple<uint, ReadOnlyMemory<char>>(value, nextSpan);
+        }
+
+        public uint GetSymbolId(ParseState state, ReadOnlySpan<char> src, int len)
+        {
+            uint nextId = (uint)state.SymbolIds.Count;
+            string key = src.Slice(0, len).ToString();
+
+            if (state.SymbolIds.TryGetValue(key, out uint existingId))
+            {
+                return existingId;
+            }
+            else
+            {
+                state.SymbolIds[key] = nextId;
+                return nextId;
+            }
+        }
+
+        public uint GenerateSymbolId(ParseState state, string baseName)
+        {
+            uint nextId = (uint)state.SymbolIds.Count;
+            string key = $"{baseName}_{nextId}";
+            state.SymbolIds[key] = nextId;
+            return nextId;
+        }
+
+        public void AddRule(ParseState state, uint ruleId, List<LLamaGrammarElement> rule)
+        {
+            while (state.Rules.Count <= ruleId)
+            {
+                state.Rules.Add(new List<LLamaGrammarElement>());
+            }
+
+            state.Rules[(int)ruleId] = rule;
+        }
+
+        public bool IsWordChar(char c)
+        {
+            return ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z') || c == '-' || ('0' <= c && c <= '9');
+        }
+
+        public Tuple<uint, ReadOnlyMemory<char>> ParseHex(ReadOnlyMemory<char> src, int size)
```

Comment:
Anywhere you're using `Tuple<A, B>` you can just use `(A, B)` instead. i.e. `(uint, ReadOnlyMemory<char>)` in this case. Same thing but nicer syntax :)